### PR TITLE
[IMP] website: polish the configurator and save info in IAP database

### PR DESCRIPTION
**Pull request by loco-odoo**
The first objective of this commit is to polish the website
configurator:
1) On page 1 of it: - The size of the text is increased.
2) On page 2 of it: - The font size inside dropdown is reduced.
                    - 30 results are displayed inside dropdown instead
of 15.
3) On page 3 of it: - The size of the text is increased.
3) On page 4: - The cards are refactored.
              - The end line is removed.
              - The "Build my website" button is moved to the left.
4) On page 5 : - It is now impossible for the user to click on a theme
preview before it is completely loaded.

5) On all pages, the size of the "Skip and start from scratch" button
has been increased.

On the second page of the website configurator, the user has to choose
an industry type. Sometimes, it happens that the user is searching for
an industry that is not known by the server. Before this commit, a
message of type "No result found" appeared and the user had to choose
an other industry among the existing ones. From this commit, the
"No result found" message is replaced by the name of the industry that
the user is writing. To choose its industry type, the user can click on
the displayed industry name or outside of the input area. This last
option calls the `_blurIndustrySelection` function and the written
industry is kept as the final industry choice. If the user chooses an
industry name that is not known by the server then the remaining of the
website configuration is done with the "abbey" industry. Moreover, this
unknown industry name as well as the language used by the user is sent
and stored to the IAP database. The goal of it is to regularly update
the industry list so that it becomes as most complete as possible.

IAP PR : https://github.com/odoo/iap-apps/pull/549
task-3062171

### CODE REVIEW

File: addons/website/static/src/client_actions/configurator/configurator.js [hunk 1]
```diff
@@ -215,6 +217,9 @@ class DescriptionScreen extends Component {
                 this.labelToId[r.label] = r.id;
             });
         } else {
+            if (terms[0].length >= this.sendIap.length) {
+                this.sendIap = terms[0];
```
Comment (rdeodoo) [inline addons/website/static/src/client_actions/configurator/configurator.js:221]: `sendIap` is not a good name, it's impossible to figure what it holds.
Maybe `unmatchTerm` ? notFoundTerm, unexistingTerm, unmatchTerm, unavailableTerm

And probably rename the python key the same: `unmatch_term` instead of `attempt`, otherwise it's too complicated to understand / figure when key has different unrelated name between the js part and the python part

File: addons/website/static/src/client_actions/configurator/configurator.js [hunk 2]
```diff
@@ -229,8 +234,15 @@ class DescriptionScreen extends Component {
      * @param {Object} ui an object with label and value properties for
      *      the selected option.
      */
-    _selectIndustry(ev, ui) {
+    async _selectIndustry(ev, ui) {
         if (this.autocompleteHasResults) {
+            if (this.sendIap.length) {
```
Comment (rdeodoo) [inline addons/website/static/src/client_actions/configurator/configurator.js:239]:
```suggestion
            if (this.sendIap) {
```
Can't have empty term, right?

File: addons/website/static/src/client_actions/configurator/configurator.js [hunk 3]
```diff
@@ -89,6 +89,8 @@ class DescriptionScreen extends Component {
         this.state = useStore();
         this.labelToId = {};
         this.autocompleteHasResults = true;
+        this.orm = useService('orm');
+        this.sendIap = '';
```
Comment (rdeodoo) [inline addons/website/static/src/client_actions/configurator/configurator.js:93]:
```suggestion
        this.sendIap = undefined;
```
No?

File: addons/website/static/src/client_actions/configurator/configurator.js [hunk 4]
```diff
@@ -229,8 +234,15 @@ class DescriptionScreen extends Component {
      * @param {Object} ui an object with label and value properties for
      *      the selected option.
      */
-    _selectIndustry(ev, ui) {
+    async _selectIndustry(ev, ui) {
```
Comment (rdeodoo) [inline addons/website/static/src/client_actions/configurator/configurator.js:237]: Is the `async` really considered in the caller (jquery autocomplete select property)?
Rather than ensuring it is, do we actually need it?
As this is about sending analytics data which we don't even do anything with the return call, we can probably not await it at all?

File: addons/website/models/website.py [hunk 5]
```diff
@@ -386,6 +386,10 @@ def configurator_skip(self):
         website = self.get_current_website()
         website.configurator_done = True
 
+    @api.model
+    def configurator_send_iap(self, attempt, final_industry_choice):
+        self._website_api_rpc('/api/website/new_industry', {'attempt': attempt, 'final_industry_choice': final_industry_choice, 'language': self.env.context.get('lang')})
```
Comment (rdeodoo) [inline addons/website/models/website.py:391]:
```suggestion
        self._website_api_rpc('/api/website/new_industry', {'attempt': attempt, 'final_industry_choice': final_industry_choice, 'lang': self.env.context.get('lang')})
```
Nitpicking, you can ignore this is you want, but we usually use `lang` shortcut everywhere.
Minor bonus (just to find one): less data sent on the network, but it's extreme nitpicking

File: addons/website/static/src/client_actions/configurator/configurator.js [hunk 6]
```diff
@@ -229,8 +234,15 @@ class DescriptionScreen extends Component {
...
         if (this.autocompleteHasResults) {
+            if (this.sendIap.length) {
+                await this.orm.call('website', 'configurator_send_iap', [], {
+                    'attempt': this.sendIap,
+                    'final_industry_choice': ui.item.label,
```
Comment (rdeodoo) [inline addons/website/static/src/client_actions/configurator/configurator.js:242]: You can probably omit the `final_` part, the shorter the better, but here it doesn't bring anything anyway as there is only one industry choice at the end: the final one

File: addons/website/static/src/client_actions/configurator/configurator.js [hunk 7]
```diff
@@ -447,17 +459,40 @@ Object.assign(FeaturesSelectionScreen, {
     template: 'website.Configurator.FeatureSelection',
 });
 
+
 class ThemeSelectionScreen extends ApplyConfiguratorScreen {
     setup() {
+        const uiService = useService('ui');
```
Comment (rdeodoo) [inline addons/website/static/src/client_actions/configurator/configurator.js:465]: move it 3 lines below next to useService('orm') and go with the same notation: this.uiService instead of const? To be consistent and not have useXXX splitted here and there

File: addons/website/static/src/client_actions/configurator/configurator.js [hunk 8]
```diff
@@ -447,17 +459,40 @@ Object.assign(FeaturesSelectionScreen, {
...
         super.setup();
 
         this.orm = useService('orm');
         this.state = useStore();
         this.themeSVGPreviews = [useRef('ThemePreview1'), useRef('ThemePreview2'), useRef('ThemePreview3')];
+        this.promiseTab = [];
```
Comment (rdeodoo) [inline addons/website/static/src/client_actions/configurator/configurator.js:471]: defs, proms, promises, please use same name as everywhere else in the code :)

Also, I'd not go with `this` here, it's not used elsewhere, except here right?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thanks for the PR.
I'll have a test + check the svg part in second pass :+1:

**Comment by loco-odoo** [comment]
@rdeodoo , as discussed last Friday, I updated the task to allow the user to choose an industry that is not known by the server.

File: addons/website/models/website.py [hunk 9]
```diff
@@ -386,6 +386,10 @@ def configurator_skip(self):
         website = self.get_current_website()
         website.configurator_done = True
 
+    @api.model
+    def configurator_send_iap(self, unmatch_term):
```
Comment (rdeodoo) [inline addons/website/models/website.py]: `configurator_send_iap` doesn't hint anything, let's rename it to `configurator_missing_industry`

**Review by rdeodoo — APPROVED** [review]
Thanks for the code and adaptation, it looks good now.

There was just a lack of consistency between the new code elements (`new` vs `wanted`, plural vs singular):
- def website_**new**_indus**tries**
- @http.route('/api/website/**new**_indus**try**'
- client.**wanted**.indus**try**
- def configurator_send_iap

But also with the rest of the existing code:
- `website.industry.theme.link` 
- `website.industry.image`
- vs `client.wanted.industry`

Let's try to uniformize everything :+1: 

I pushed this diff:
- Community [diff](https://github.com/odoo/odoo/compare/d279228dc544830d2d4de216a2dcb3332d741c00..3ca6f093341fec0eeec651d0a8aac2154da34b72)
- Iap [diff](https://github.com/odoo/iap-apps/compare/f54a3650691c3197266865933e2086f770090e2e..1856c719067e9678f694a8bc695f6c6f731927bf)

I'll have a final look now

File: addons/website/static/src/client_actions/configurator/configurator.js [hunk 10]
```diff
@@ -451,13 +417,35 @@ class ThemeSelectionScreen extends ApplyConfiguratorScreen {
...
+                            reject(imgEl);
+                        }, {once: true});
+                    }));
+                }
+                $(this.themeSVGPreviews[idx].el).append(svgEl);
+            });
+            // When all the images inside the svgs are loaded then remove the
+            // loading effect.
+            Promise.all(proms).then(() => {
+                this.uiService.unblock();
```
Comment (rdeodoo) [inline addons/website/static/src/client_actions/configurator/configurator.js:448]: This is problematic, it will never be executed if one of the promise is rejected.
That's what I already faced in local, for some reason I got a key error on a theme image (probably wrong theme version) which made a single image out of all of the images not load (as it was not found / 404).
The loading would never stop and spin forever as no code would release it.

Need to replace it by `allSettled` I guess?

File: addons/website/static/src/client_actions/configurator/configurator.xml [hunk 11]
```diff
@@ -149,33 +149,35 @@
...
                                     </div>
                                 </div>
                             </div>
                         </t>
                     </div>
                 </div>
-                <div class="text-end border-top pt-2">
-                    <button class="btn btn-primary btn-lg ms-3" t-on-click="buildWebsite">Build my website</button>
+                <div class="pt-2">
+                    <button class="btn btn-primary btn-lg btn-build" t-on-click="buildWebsite">Build my website</button>
```
Comment (rdeodoo) [inline addons/website/static/src/client_actions/configurator/configurator.xml:180]: btn-build ?

File: addons/website/static/src/client_actions/configurator/configurator.js [hunk 12]
```diff
@@ -215,42 +183,33 @@ class DescriptionScreen extends Component {
                 this.labelToId[r.label] = r.id;
             });
         } else {
-            labels = [_t("No result found, broaden your search.")];
+            labels = [_t(request.term)];
```
[the remark this answers is not in the document]
Comment (loco-odoo) [reply addons/website/static/src/client_actions/configurator/configurator.js]: Just to be sure, this `_t()` was about translating its input right?

File: addons/website/static/src/client_actions/configurator/configurator.xml (hunk 11 above)
Comment (loco-odoo) [reply addons/website/static/src/client_actions/configurator/configurator.xml]: > btn-build ?

Sorry, I think was for a test :smile: I just removed it :+1:

File: addons/website/static/src/client_actions/configurator/configurator.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/static/src/client_actions/configurator/configurator.js]: It will extract the given term to the `.pot` file which then will generate traduction requests on transifex website which then will generate the according `.po` file with translation in Odoo.
But here obviously it won't do anything as there is nothing to be sent on transifex for translation.

[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/static/src/client_actions/configurator/configurator.js]: Don't hesitate to ask whenever you find something that you don't understand or have doubt about, even if it's 10 times a day.
Here you modified that line and should have been curious about it.
Ask and get answer, there is no harm. We are here for that and just that: spread knowledge. It will increase your learning curve x10 ;-)

**Comment by loco-odoo** [comment]
@rdeodoo, I just pushed a slight modif : `_t` was not used anymore in the file so I removed it from the import.

**Comment by rdeodoo** [comment]
Your branch was 740 commits behind, try to rebase it from time to time. Here the runbot was red because of that (most likely) but even with a green runbot, it's not ideal as you/we are testing the diff/feature on an very outdated branch, meaning that once rebased many changes will be applied and could lead to your behavior to not work anymore etc.

Also, rebasing often will make it easier for you to fix conflicts, as you won't have a lot to fix in one go.

tl;dr: rebase often your branch :+1: (Here I just did it)

Bonus point: try to push a theme branch so we can access the design build on your runbot without having to build it manually (and wait 15 minutes) :+1: 
It's important when doing changes which indirectly implies themes.

**Comment by rdeodoo** [comment]
IAP part has been [merged](https://github.com/odoo/iap-apps/commit/ee723f202da8183536d1925965c29f714e9e41a4) but not yet deployed.
Once it's deployed, this one can be shipped :+1: 
Probably by early next week, @lobaudoux will let us know I guess :)

**Review by rdeodoo — APPROVED** [review]
Marvelous, thank you.

@robodoo r+
